src/ir/auth_logic/ast.h

```diff
@@ -139,6 +139,46 @@ class Assertion {
   AssertionVariantType value_;
 };
 
+enum class ConstantType { NumberType, PrincipalType };
+
+// Relation declaration have types are of 3 forms
+// 1. NumberType
+// 2. PrincipalType
+// 3. User defined CustomType (string to store name of the type)
+class ArguementType {
+ public:
+  enum class Kind { kNumber, kPrincipal, kCustom };
+  explicit ArguementType(Kind kind, std::string name)
+      : kind_(std::move(kind)), name_(std::move(name)) {}
+  explicit ArguementType(Kind kind) : kind_(std::move(kind)) {}
```

Comment:
Why is this needed? Don't we always want to have a name? It looks like you expect name only if kind is `CustomType`? An alternative way to do this with static factory methods and simply hide the constructor:

```C++
class ArgumentType {
public:
  // ...
  static ArgumentType MakePrincipalType() { return ArgumentType(Kind::kPrincipal, ""); }
  static ArgumentType MakeNumberType() { return ArgumentType(Kind::kNumber, ""); }
  static ArgumentType MakeCustomType(absl::string_view name) { return ArgumentType(Kind::kCustom, name); }

  ...
private:
     explicit ArgumentType(Kind kind, absl::string_view name) : kind_(std::move(kind)), name_(name) {}
}
```


```

Reply:
+1. This has the additional benefit of being more type-safe, in the sense that you are disallowing the bad pattern of making a `CustomType` without a name or a `NumberType` with a name.

Reply:
I made changes to ArgumentType Node to make it look simpler. The reason I left out the name is that, for NumberType and PrincipalType, it is understood that they correspond to 'Number' and 'Principal' (Handle this in ast_construction). Whereas for customType, we need to explicitly provide the name for type. As this does not look like a good pattern to have, I changed to having name for all types. PTAL.